Alpha release candidate (#248)
Agoric/documentation: main/zoe/api/zoe.md

```js
const issuerKeywordRecord = await E(zoe).getIssuers(instance);
```
## E(zoe).getTerms(instance)
- `instance` `{Promise<Instance>}`
- Returns: `{Object}` 

Returns the terms of the `instance` argument, including its `issuers`, `brands` and any 
custom terms.
```js
const terms = await E(zoe).getTerms(instance);
```

## E(zoe).getPublicFacet(instance)
- `instance` `{Instance}`
- Returns: `{Promise<PublicFacet>}` 

A contract instance's `publicFacet` is an object available via Zoe to anyone knowing that `instance`. 
You use it for general queries and actions, such as getting a current price or creating public invitations.
Since a facet is defined just as any other object, you add methods to the `publicFacet` just like you would
any object.

Returns a `publicFacet` containing the public facet defined for `instance`.
```js
const ticketSalesPublicFacet = await E(zoe).getPublicFacet(sellItemsInstance);
```
## E(zoe).getInvitationIssuer()
- Returns `{Promise<Issuer>}`

Zoe has a single `invitationIssuer` for its entire
lifetime. By having a reference to Zoe, a user can get the `invitationIssuer` and 
validate any `invitation` they receive

> from someone else by calling `E(invitationIssuer).claim()` with the 

Author: Hmm. I'm thinking what I meant to convey here was that claim handles validation while claiming it, and less "here's an way of validating it" period. Would the following work?

By having a reference to Zoe, a user can get the `invitationIssuer`. This lets them claim any
invitation they receive from someone else by calling `E(invitationIssuer).claim()` with the
untrusted invitation as the argument. During the claiming process, the `invitationIssuer` validates
the invitation.

> from someone else by calling `E(invitationIssuer).claim()` with the 

Reviewer: yep. that's fine.

Returns an `installation` object. 

```js
// bundleSource takes source code files and 
// bundles them together in the format install expects.
import bundleSource from '@agoric/bundle-source';
const bundle = await bundleSource(pathResolve(`./src/contract.js`));
const installationP = await E(zoe).install(bundle);
```

## E(zoe).getInstance(invitation)
- `invitation` `{Invitation}`
- Returns: `{Promise<Instance>}`

Returns a `Promise` for the contract `instance` the `invitation` is part of.
```js
const instance = await E(zoe).getInstance(invitation);
```
      
## E(zoe).getInstallation(invitation)
- `invitation` `{Invitation}`
- Returns: `{Promise<Installation>}`

Returns a `Promise` for the contract `installation` the `invitation`'s contract `instance` uses.
```js
const installation = await E(zoe).getInstallation(invitation);
```
## E(zoe).startInstance(installation, issuerKeywordRecord, terms)
- `installation` `{Installation}`
- `issuerKeywordRecord` `{IssuerKeywordRecord}`
- `terms` `{Object}`
- Returns: `{Promise<StartInstanceResult>}`

Create an `instance` of the installed smart contract (specified by 
the `installation` argument). You must also specify the 

> instance's `issuerKeywordRecord` and `terms` for the contract 
> (as key-value pairs). 

Reviewer: ```suggestion
instance's `issuerKeywordRecord` (as key-value pairs) and `terms` for the contract. 
```

The `issuerKeywordRecord` is a record mapping string names (keywords) 
to `issuers`, such as `{ Asset: quatlooIssuer}`. Keywords must begin 
with a capital letter and must be ASCII. Parties to the contract will 
use the keywords to index their proposal and their payments. 

`terms` are values used by this contract instance, such as the 
number of bids an auction will wait for before closing. These values may
be different for different instances of the same contract, but the contract
defines what variables need their values passed in as `terms`. 

It returns a `promise` for a `StartInstanceResult` object. The object consists of:
- `creatorFacet` `{any}`
- `publicFacet` `{any}`
- `instance` `{Instance}`
- `creatorInvitation` `{Payment | undefined}`

> A `publicFacet` and a `creatorFacet` are objects available via Zoe to anyone knowing 

Reviewer: ```suggestion
A `publicFacet` is an object available via Zoe to anyone knowing 
```

The `creatorFacet` is only available in this return value. The contract designer should use it to encapsulate things that the contract runner might not want to share, or might want to control the distribution of. The party who starts the contract should carefully consider the impact before sharing access to the creatorFace.

the `instance` they are associated with. The `publicFacet` is used for general queries 
and actions, such as getting a current price or creating public invitations. Since a 

> facet is defined just as any other object, you add methods to them just like you would any object.

Reviewer: ```suggestion
facet is defined just as any other object, the contract developer can add methods to them just like they would any object.
```

`creatorInvitation` is an invitation that the contract instance creator can use. 
It is usually used in contracts where the creator immediately sells 
something (auctions, swaps, etc.), so it's helpful for the creator to have 
an invitation to escrow and sell goods.

```js
const issuerKeywordRecord = { 
  'Asset' : moolaIssuer, 
  'Price' : quatlooIssuer 
};
const terms = { numBids: 3 };
const { invite, instanceRecord } = await E(zoe).startInstance(
  creatorFacet, publicFacet, creatorInvitation);
```
## E(Zoe).offer(invitation, proposal, paymentKeywordRecord)
- `invitation` `{Invitation|Promise<Invitation>}`
- `proposal` `{Proposal}`
- `paymentKeywordRecord` `{PaymentKeywordRecord}`
- Returns: `{Promise<UserSeat>}`

Used to redeem the `invitation` provided as the first argument.

> To redeem an `invitation`, a user normally provides a `proposal` (their
> rules for an offer) as well as `payments` to be escrowed by Zoe.  If

Reviewer: ```suggestion
To redeem an `invitation`, a user normally provides a `proposal` (the
rules under which they want to exercise the offer) as well as `payments` to be escrowed by Zoe.  If
```

either the `proposal `or `payments` are empty, indicate this by
omitting that argument or passing `undefined`, instead of passing an
empty record.

The `proposal` has three parts: 
- `want`: An object with keywords as keys and `amounts` as values.
- `give`: An object with keywords as keys and `amounts` as values.
- `exit`: Specifies the payout-liveness policy Zoe can guarantee for the offer.

`paymentKeywordRecord` is a record with keywords as keys, with
 values of the actual `payments` to be escrowed. A `payment` is

>  expected for every rule under `give`.

Reviewer: ```suggestion
 expected for every entry under `give`.
```

>  `offer()` returns a `promise` for a `userSeat`. 

Reviewer: we need a description of the API for userSeat.

Author: It's coming shortly. After this lands, I'll do a hopefully quick v2 that incorporates both objects like UserSeat and ZCFSeat  and the current contents of the Records doc page (updated to Alpha).

::: tip Zoe is accessed asynchronously
The Zoe service is accessed asynchronously, using a standards-track library extension
to JavaScript that uses promises as remote references. In code, the Zoe service instance 
is referred to via `zoe`, which only supports asynchronous invocation. Operations are 
invoked asynchronously using the [`E` helper for async messaging](https://github.com/tc39/proposal-eventual-send#e-and-esendonly-convenience-proxies). 
All such operations immediately return a promise for their result. That may eventually fulfill to a local value, or to a `Presence` for another remote object (e.g. in another contract or service, running on another chain, etc.). Async messages can be sent using `E` with either promises or presences. 

For more information about using `E`, see the section on it in [Agoric's JavaScript Distributed Programming Guide](https://agoric.com/documentation/distributed-programming.html). 
:::

**tyg todo: Does cleanProposal.js need to be documented?**

> **tyg todo: Where should the "start" method(?) be documented? i.e.**

Author: Am assuming the first comment applies to "cleanProposal.js" and the second to "start"?

> **tyg todo: Where should the "start" method(?) be documented? i.e.**

Reviewer 2: We don't need to document cleanProposal at all, and both of these comments apply to `start`

```js
const start = zcf => {
  ...
  // your code here
  return harden({ creatorFacet, creatorInvitation, publicFacet });
}
harden(start);
export { start };
```

## E(zoe).getBrands(instance)
- `instance` `{Instance}`
- Returns: `{BrandKeywordRecord}`

> **tyg todo: Should all the "getFoo()" methods return** {Promise<Record>}? 

Reviewer 2: Sure, I think it's good for people to expect a promise if we document the calls like `E(something)`

> **tyg todo: And related, are they always called with "await"?**

Reviewer 2: `await` is one way of waiting for a promise to resolve and getting the value to which it resolves. You can also do `promise.then`. this means that you never use `await` with non-promises. We have more best practices for when we internally can use `await` safely, but I don't think there's any need for that to go in the external documentation at this point. We can use `await` in our code examples.

```
## E(zoe).getTerms(instance)
- `instance` `{Instance}`
- Returns: `{Object}` consisting of key:value pairs **tyg todo: Is there a better way to state this here?**
**tyg todo: Not sure if this is right, or if it should be the same as zcf.getTerms()**
Terms let a contract instance creator further customize the contract operations, 
as enabled by the contract code. Contract terms can be expressed without specific values. 
For example, an auction contract may define minimum bid and minimum raise variables and 
use them in its code, but the variables are not given values in the code. 
Instead, the terms given at contract instantiation provide the variables’ values.
```js
const terms = await E(zoe).getTerms(instance);
```

## E(zoe).getPublicFacet(instance)
- `instance` `{Instance}`
- Returns: `{PublicFacet}`  **tyg todo: Is this right?**

A contract instance's `publicFacet` is an object available via Zoe to anyone knowing that `instance`. 
You use it for general queries and actions, such as getting a current price or creating public invitations.
**tyg todo: I'd really like to beef this up a bit, including what needs and should be done for a contract

> developer to create a publicFacet (or pointer to same) and details of how to use it. Is it just publicFacet.getCurrentPrice() or similar

Author: Right, but how do you attach methods (and objects?) to the public facet?

> developer to create a publicFacet (or pointer to same) and details of how to use it. Is it just publicFacet.getCurrentPrice() or similar

Reviewer 2: The publicFacet is an object. You define it like any other object, within the contract code, and return it. You cannot attach methods or add new properties after returning it from the `start` function. I think our sample contracts are good examples of how publicFacets are created.

depending on its definition?**

Returns a `IssuerKeywordRecord` containing the public facet defined in the argument contract `instance`.
```js
const ticketSalesPublicFacet = await E(zoe).getPublicFacet(sellItemsInstance);
```
## E(zoe).getInvitationIssuer()
- Returns `{Issuer}`

Zoe has a single `invitationIssuer` for its entire
lifetime. By having a reference to Zoe, a user can get the `invitationIssuer` and 
validate any `invitation` they receive
from someone else by calling `invitationIssuer.claim()` with the 
untrusted invitation as the argument.

The `mint` associated with the `invitationIssuer`
create `invitations` in the form of ERTP `payments` that represent the right to interact with
a smart contract in particular ways.

> **tyg todo: Need more info on how to use this, including just making invitations and

Author: How does one add an ability to a facet?

what the .claim and .getAmountOf methods do/are for**
```js
const invitationIssuer = await E(zoe).getInvitationIssuer();
// Here a user, Bob, has received an untrusted invitation from Alice.
// Bob uses the trusted `invitationIssuer` from Zoe to
// transform the untrusted invitation to a trusted one
const invitation = await invitationIssuer.claim(untrustedInvitation);
const invitationValue = await E(zoe).getInvitationDetails(invitation);

> const { value: invitationValue } = await E(invitationIssuer).getAmountOf(

Author: Not sure I'm clear on this; need to discuss invitationIssuer.

        invitation);
```
## E(zoe).getInvitationDetails(invitation)
- `invitation` `{Invitation}`
- Returns `{Object}`

Takes an `invitation` as an argument and returns an object containing the following 
details about the argument:
- `installation` `{Installation}`: The contract's Zoe installation.
- `instance` `{Instance}`: The contract instance this invitation is in.
- `invitationHandle` `{Handle}`: A reference used to refer to this invitation.
- `description` `{String}`: A description of the invitation, covering what a recipient needs to know about the contract and participating in it, as well as anything about the invitation, such as an expiration date after which the invitation is invalid.
```js
const invitation = await invitationIssuer.claim(untrustedInvitation);
const invitationValue = await E(zoe).getInvitationDetails(invitation);
```

## E(zoe).install(bundle)
- `bundle` `{SourceBundle}`
- Returns: `{Promise<Installation>}`

Takes bundled source code for a Zoe contract as an argument and installs the code on Zoe.

> Returns an `installation` object. **tyg todo: Cover how to get code bundled (or link to same)**

Reviewer 2: Yes, I'm answering this exact question. Take a look at any of our tests for the contracts. You will see exactly how it is done. For instance:

```js
import bundleSource from '@agoric/bundle-source';
const filePath = `atomicSwap`; // whatever the filepath is
const bundle = await bundleSource(filePath);
const installationP = E(zoe).install(bundle);
```
That is all you do.

To redeem (sometimes called "exercise") an `invitation`, a user normally provides a `proposal` (their
rules for the offer) as well as `payments` to be escrowed by Zoe.  If
either the `proposal `or `payments` are empty, indicate this by
omitting that argument or passing `undefined`, instead of passing an
empty record.

The `proposal` has three parts: 
- `want`: An object with keywords as keys and amounts as values.
- `give`: An object with keywords as keys and amounts as values.
- `exit`: Specifies the payout-liveness policy Zoe can guarantee for the offer.

`paymentKeywordRecord` is a record with keywords as keys, with
 values of the actual `payments` to be escrowed. A `payment` is
 expected for every rule under `give`.
 
 `offer()` returns a `promise` for a `userSeat`. See the Objects section for its
 description.
 ```js
 **tyg todo: Need good source code**
 ```

---------------------------
**tyg todo: As mentioned in the standup, I think we now need to cover
API objects in the doc. I'm undecided if they should be one page with methods,
on have their own pages. Thoughts? For the time being, I'm including them on the
same page with a h1 heading**

> # Objects

Author: I've got two concerns about this:

1) Findability. If objects are described within a method's documentation, the indexing is going to be on the method. If the object doesn't have effectively the same name as the method, it'll be hard to find. If an object is used at all outside of its immediate creation by a method, it should have its own, indexable and findable, entry.

2) Length. Some of these objects, particularly when I add on method documentation for methods on them, are already going to be pretty long in docs. Adding them to method entries will make for several very long entries. On the other hand, people are mostly likely to use the reference one (method) entry at a time, so this may not be a significant issue. 

If we're going to intermingle method and object writeups, I'd prefer it be of the form:

Method Entry with ## title to index from   

Related to above method Object Entry with ## to index from
Additional related to above method Object Entry with ## to index from

Next Method ## Entry.

> # Objects

Reviewer 2: As discussed elsewhere, we are not doing an object section, and are instead moving the objects into the various sections in which they are introduced/used, but are collapsing the object part such that it doesn't take up too much room. Additionally, h1, h2, and h3, all show up in the sidebar, so we can use `###` headers for the objects and they should still show up in the sidebar.

The Zoe API defines and uses several object types. This **tyg todo section/page** 
shows their definitions and discusses their uses.

## Installation Object

`installation` objects represent stored in Zoe contract source code. 
They have one method and no properties.

- installation.getBundle()
  - Returns: `{SourceBundle}`
  - Gets and returns the bundled source code for its associated contract.

## StartInstanceResult Object
`startInstance()` returns a `promise` for a `StartInstanceResult` object. It has four properties: of:
- `creatorFacet` `{any}`
- `publicFacet` `{any}`
- `instance` `{Instance}`
- `creatorInvitation `{Payment | undefined}`
 

## InvitationDetails Object

`invitationDetails` objects have four properties:
- `installation` `{Installation}`: The contract's Zoe installation.
- `instance` `{Instance}`: The contract instance this `invitation` is in.
- `invitationHandle` `{Handle}: A reference that refers to this `invitation`.

> - `description` `{String}`: A description of the `invitation`, covering what 

Reviewer 2: The other explanation got resolved without this getting fixed - the description doesn't have any information in it other than serving as a name of the invitation. The name can be used to find the particular part of the contract code that created the invitation.
